### PR TITLE
Feature battery refactor

### PR DESCRIPTION
**Pull request by ssankalp26**

### CODE REVIEW

File: RCAIDE/Framework/Mission/Segments/Ground/Battery_Recharge.py [hunk 1]
```diff
@@ -49,8 +49,6 @@ def __defaults__(self):
         #   User Inputs
         # --------------------------------------------------------------
         self.altitude                      = None
-        self.time                          = 1.0 * Units.seconds
-        self.current                       = 0
         self.overcharge_contingency        = 1.10
```
Comment (ssankalp26) [inline RCAIDE/Framework/Mission/Segments/Ground/Battery_Recharge.py:52]: add self.time

File: RCAIDE/Framework/Networks/Electric.py [hunk 2]
```diff
@@ -113,82 +112,79 @@ def evaluate(self,state,center_of_gravity):
...
+                for battery in  bus.battery_modules:                   
+                    if bus.identical_batteries == False:
+                        # run analysis  
+                        stored_results_flag, stored_battery_tag =  battery.energy_calc(state,bus,coolant_lines, t_idx, delta_t, recharging_flag)
+                    else:             
+                        if stored_results_flag == False: 
+                            # run propulsor analysis 
+                            stored_results_flag, stored_battery_tag  =  battery.energy_calc(state,bus,coolant_lines, t_idx, delta_t, recharging_flag)
+                        else:
+                            # use previous propulsor results
```
Comment (ssankalp26) [inline RCAIDE/Framework/Networks/Electric.py:174]: change comment

File: RCAIDE/Framework/Networks/Electric.py [hunk 3]
```diff
@@ -347,5 +343,7 @@ def add_unknowns_and_residuals_to_segment(self, segment):
         segment.process.iterate.residuals.network           = self.residuals        
 
         return segment
+    __call__ = evaluate 
+def run_energy_calc(battery, state, bus, coolant_lines, t_idx, delta_t, recharging_flag):
```
Comment (ssankalp26) [inline RCAIDE/Framework/Networks/Electric.py:347]: delete this

File: RCAIDE/Library/Components/Energy/Distributors/Electrical_Bus.py [hunk 4]
```diff
@@ -28,15 +30,30 @@ def __defaults__(self):
...
+        self.battery_module_electric_configuration = None 
+        
+    def append_operating_conditions(self, segment):
+        append_bus_conditions(self, segment)
+        return
+    
+    def compute_distributor_conditions(self,state):
+        compute_bus_conditions(self, state)
+        return 
+
```
Comment (ssankalp26) [inline RCAIDE/Library/Components/Energy/Distributors/Electrical_Bus.py:57]: delete extra lines

File: RCAIDE/Library/Components/Energy/Sources/Battery_Modules/Generic_Battery_Module.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        Source:
+            None
+        """
+      
+       
+        self.energy_density                         = 0.0
+        self.current_energy                         = 0.0
+        self.current_capacitor_charge               = 0.0
+        
+        self.capacity                           = 0.0
```
Comment (ssankalp26) [inline RCAIDE/Library/Components/Energy/Sources/Battery_Modules/Generic_Battery_Module.py:37]: align

File: RCAIDE/Library/Components/Thermal_Management/Heat_Exchangers/Cross_Flow_Heat_Exchanger.py [hunk 6]
```diff
@@ -44,7 +44,7 @@ def __defaults__(self):
         self.tag                                                    = 'cross_flow_heat_exchanger'
         self.coolant                                                = Glycol_Water() 
         self.air                                                    = Air() 
-
+        self.design_heat_removed   =  0.0
```
Comment (ssankalp26) [inline RCAIDE/Library/Components/Thermal_Management/Heat_Exchangers/Cross_Flow_Heat_Exchanger.py:47]: align

File: RCAIDE/Library/Components/Thermal_Management/Reservoirs/Reservoir.py [hunk 7]
```diff
@@ -39,9 +39,9 @@ def __defaults__(self):
         self.tag                          = 'coolant_reservoir'
         self.material                     = Polyetherimide()
         self.coolant                      = Glycol_Water()
-        self.length                       = 0.5                                      # [m]
-        self.width                        = 0.5                                      # [m]
-        self.height                       = 0.5                                      # [m]
+        self.length                       = 5                                     # [m]
+        self.width                        = 5                                     # [m]
+        self.height                       = 5                                     # [m]
```
Comment (ssankalp26) [inline RCAIDE/Library/Components/Thermal_Management/Reservoirs/Reservoir.py:44]: reduce to normal dimensions and check if things break

File: RCAIDE/Library/Methods/Energy/Distributors/Electrical_Bus/append_bus_conditions.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+# ----------------------------------------------------------------------------------------------------------------------
+#  IMPORT
+# ----------------------------------------------------------------------------------------------------------------------
+# RCAIDE imports  
+from RCAIDE.Framework.Mission.Common     import   Conditions
+
+# ----------------------------------------------------------------------------------------------------------------------
+#  METHODS
+# ---------------------------------------------------------------------------------------------------------------------- 
+## @ingroup Methods-Energy-Sources-Battery
```
Comment (ssankalp26) [inline RCAIDE/Library/Methods/Energy/Distributors/Electrical_Bus/append_bus_conditions.py:14]: delete

File: RCAIDE/Library/Methods/Energy/Distributors/Electrical_Bus/append_bus_conditions.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        None
+        """
+    ones_row                                                       = segment.state.ones_row
+   
+    segment.state.conditions.energy[bus.tag]                       = Conditions()
+    segment.state.conditions.energy[bus.tag].battery_modules       = Conditions()
+    segment.state.conditions.energy[bus.tag].power_draw            = 0 * ones_row(1)
+    segment.state.conditions.energy[bus.tag].SOC                   = 0 * ones_row(1) 
+    segment.state.conditions.energy[bus.tag].current_draw          = 0 * ones_row(1)
+    segment.state.conditions.energy[bus.tag].charging_current      =  0 * ones_row(1)
```
Comment (ssankalp26) [inline RCAIDE/Library/Methods/Energy/Distributors/Electrical_Bus/append_bus_conditions.py:38]: align

File: RCAIDE/Library/Methods/Energy/Distributors/Electrical_Bus/compute_bus_conditions.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    
+    bus_conditions                         = state.conditions.energy[bus.tag]
+    
+    for battery_module in  bus.battery_modules:
+        bm_conditions                          =  bus_conditions.battery_modules[battery_module.tag]
+        bus_conditions.voltage_open_circuit    += bm_conditions.voltage_open_circuit
+        bus_conditions.voltage_under_load      += bm_conditions.voltage_under_load 
+        bus_conditions.heat_energy_generated   += bm_conditions.heat_energy_generated
+        bus_conditions.temperature             += bm_conditions.temperature
+        bus_conditions.energy                  += bm_conditions.energy
```
Comment (ssankalp26) [inline RCAIDE/Library/Methods/Energy/Distributors/Electrical_Bus/compute_bus_conditions.py:23]: space

File: RCAIDE/Library/Methods/Energy/Distributors/Electrical_Bus/compute_bus_conditions.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+# Created: Sep 2024, S. Shekar
+
+# ----------------------------------------------------------------------------------------------------------------------
+#  IMPORT
+# ----------------------------------------------------------------------------------------------------------------------
+
+## @ingroup Methods-Energy-Sources-Battery 
+def compute_bus_conditions(bus,state): 
+    """  
+    """
```
Comment (ssankalp26) [inline RCAIDE/Library/Methods/Energy/Distributors/Electrical_Bus/compute_bus_conditions.py:13]: function header

File: RCAIDE/Library/Methods/Energy/Sources/Batteries/Common/append_battery_conditions.py [hunk 12]
```diff
@@ -68,71 +68,82 @@ def append_battery_conditions(battery,segment,bus):
...
-        segment.state.conditions.energy.recharging  = False 
-        segment.state.unknowns['discharge']          =  0* ones_row(1)  
-        segment.state.residuals.network['discharge'] =  0* ones_row(1) 
-        bus_results[battery.tag].pack.charging_current = -segment.current * ones_row(1)
+        #bus_results.battery_modules[battery.tag].charging_current = segment.current * ones_row(1)  
+    #elif type(segment) == RCAIDE.Framework.Mission.Segments.Ground.Battery_Discharge:
+        #segment.state.conditions.energy.recharging  = False 
+        #segment.state.unknowns['discharge']          =  0* ones_row(1)  
+        #segment.state.residuals.network['discharge'] =  0* ones_row(1) 
+        #bus_results.battery_modules[battery.tag].charging_current = -segment.current * ones_row(1)
```
Comment (ssankalp26) [inline RCAIDE/Library/Methods/Energy/Sources/Batteries/Common/append_battery_conditions.py:116]: delete

File: RCAIDE/Library/Methods/Energy/Sources/Batteries/Common/append_battery_conditions.py [hunk 13]
```diff
@@ -157,16 +168,16 @@ def append_battery_segment_conditions(battery, bus, conditions, segment):
...
         else:                   
             battery_conditions.battery_discharge_flag           = True             
-        battery_conditions.pack.maximum_initial_energy          = battery_initials.pack.maximum_initial_energy 
-        battery_conditions.pack.energy[:,0]                     = battery_initials.pack.energy[-1,0]
-        battery_conditions.pack.temperature[:,0]                = battery_initials.pack.temperature[-1,0]
+        battery_conditions.maximum_initial_energy          = battery_initials.maximum_initial_energy 
+        battery_conditions.energy[:,0]                     = battery_initials.energy[-1,0]
+        battery_conditions.temperature[:,0]                = battery_initials.temperature[-1,0]
         battery_conditions.cell.temperature[:,0]                = battery_initials.cell.temperature[-1,0]
         battery_conditions.cell.cycle_in_day                    = battery_initials.cell.cycle_in_day
```
Comment (ssankalp26) [inline RCAIDE/Library/Methods/Energy/Sources/Batteries/Common/append_battery_conditions.py:182]: align

File: RCAIDE/Library/Methods/Energy/Sources/Batteries/Common/append_battery_conditions.py [hunk 14]
```diff
@@ -68,71 +68,82 @@ def append_battery_conditions(battery,segment,bus):
...
-    bus_results[battery.tag].cell.heat_energy_generated    = 0 * ones_row(1)    
-    bus_results[battery.tag].cell.power                    = 0 * ones_row(1)         
-    bus_results[battery.tag].cell.energy                   = 0 * ones_row(1)         
-    bus_results[battery.tag].cell.voltage_under_load       = 0 * ones_row(1)         
-    bus_results[battery.tag].cell.voltage_open_circuit     = 0 * ones_row(1)        
-    bus_results[battery.tag].cell.current                  = 0 * ones_row(1)  
-    bus_results[battery.tag].cell.cycle_in_day             = 0
-    bus_results[battery.tag].cell.resistance_growth_factor = 1.
-    bus_results[battery.tag].cell.capacity_fade_factor     = 1. 
+    bus_results                                              = segment.state.conditions.energy[bus.tag]
```
Comment (ssankalp26) [inline RCAIDE/Library/Methods/Energy/Sources/Batteries/Common/append_battery_conditions.py:71]: align

File: RCAIDE/Library/Methods/Energy/Sources/Batteries/Common/initialize_from_energy_and_power.py [hunk 15]
```diff
@@ -42,17 +42,17 @@ def initialize_from_energy_and_power(battery, energy, power, max='hard'):
...
         
     else:
         mass=np.maximum(energy_mass, power_mass)
 
-    battery.mass_properties.mass = mass
-    battery.pack.maximum_energy  = battery.specific_energy*mass
-    battery.pack.maximum_power       = battery.specific_power*mass
+    battery.mass_properties.mass   = mass
+    battery.maximum_energy  = battery.cell.specific_energy*mass
+    battery.maximum_power   = battery.cell.specific_power*mass
```
Comment (ssankalp26) [inline RCAIDE/Library/Methods/Energy/Sources/Batteries/Common/initialize_from_energy_and_power.py:56]: align

File: RCAIDE/Library/Methods/Energy/Sources/Batteries/Common/initialize_from_mass.py [hunk 16]
```diff
@@ -36,14 +36,15 @@ def initialize_from_mass(battery,module_weight_factor = 1.42 ):
...
\ No newline at end of file
+    battery.maximum_energy                    = mass*battery.specific_energy  
+    battery.maximum_power                     = mass*battery.specific_power
+    battery.initial_maximum_energy            = battery.maximum_energy    
+    battery.electrical_configuration.series   = n_series
+    battery.electrical_configuration.parallel = n_parallel 
+    battery.electrical_configuration.total    = n_parallel*n_series      
+    #battery.charging_voltage                       = battery.cell.charging_voltage * battery.pack.electrical_configuration.series     
+    #battery.charging_current                       = battery.cell.charging_current * battery.pack.electrical_configuration.parallel
+    # Can I delete this from REGRESSIONS??????
```
Comment (ssankalp26) [inline RCAIDE/Library/Methods/Energy/Sources/Batteries/Common/initialize_from_mass.py:50]: @mclarke2

File: RCAIDE/Library/Methods/Thermal_Management/Batteries/Liquid_Cooled_Wavy_Channel/wavy_channel_geometry_setup.py [hunk 17]
```diff
@@ -29,11 +30,14 @@ def wavy_channel_geometry_setup(HAS,battery):
...
-    bus.batteries.append(battery)
-    net.busses.append(bus) 
+    net                                                       = RCAIDE.Framework.Networks.Electric()
+    bus                                                       = RCAIDE.Library.Components.Energy.Distributors.Electrical_Bus()
+    bus.battery_modules.append(battery)
+    coolant_line                                              = RCAIDE.Library.Components.Energy.Distributors.Coolant_Line(bus)     
+    coolant_line.battery_modules[battery.tag].thermal_management_system       = Data() 
+    coolant_line.battery_modules[battery.tag].thermal_management_system.heat_acquisition_system = HAS 
+    net.busses.append(bus)
+    net.coolant_lines.append(coolant_line)
```
Comment (ssankalp26) [inline RCAIDE/Library/Methods/Thermal_Management/Batteries/Liquid_Cooled_Wavy_Channel/wavy_channel_geometry_setup.py:40]: block text

File: RCAIDE/Library/Methods/Thermal_Management/Batteries/Liquid_Cooled_Wavy_Channel/wavy_channel_sizing_setup.py [hunk 18]
```diff
@@ -77,9 +77,9 @@ def modify_wavy_channel_HAS(nexus):
     h_cell    = battery.cell.height                      
     A_cell    = np.pi*d_cell*h_cell    
     T_bat     = has_opt.design_battery_operating_temperature
-    Q_pack    = has_opt.design_heat_removed 
-    N_mod     = battery.pack.number_of_modules
-    Q_module  = Q_pack/N_mod 
+    Q_module    = has_opt.design_heat_removed 
+    #N_mod     = battery.pack.number_of_modules
+    #Q_module  = Q_pack/N_mod
```
Comment (ssankalp26) [inline RCAIDE/Library/Methods/Thermal_Management/Batteries/Liquid_Cooled_Wavy_Channel/wavy_channel_sizing_setup.py:82]: delete

File: RCAIDE/Library/Methods/Thermal_Management/Heat_Exchangers/Cross_Flow_Heat_Exchanger/cross_flow_heat_exchanger_geometry_setup.py [hunk 19]
```diff
@@ -26,16 +27,22 @@ def cross_flow_heat_exchanger_geometry_setup(HEX):
...
-    vehicle.append_energy_network(net)
+    vehicle                                                                  = RCAIDE.Vehicle()  
+    net                                                                      = RCAIDE.Framework.Networks.Electric()
+    #bus                                                                      = RCAIDE.Library.Components.Energy.Distributors.Electrical_Bus()
+    #bus.battery_modules.append(battery)                
+    coolant_line                                                             = deepcopy(coolant_line_base) 
+    # Just to create an empty container for the Heat Exchanger
+    RCAIDE.Library.Components.Thermal_Management.Batteries.Liquid_Cooled_Wavy_Channel(coolant_line)
+    
+    HEX.coolant_temperature_of_hot_fluid                  = 313
```
Comment (ssankalp26) [inline RCAIDE/Library/Methods/Thermal_Management/Heat_Exchangers/Cross_Flow_Heat_Exchanger/cross_flow_heat_exchanger_geometry_setup.py:38]: align

File: RCAIDE/Library/Methods/Thermal_Management/Heat_Exchangers/Cross_Flow_Heat_Exchanger/cross_flow_heat_exchanger_geometry_setup.py [hunk 20]
```diff
@@ -26,16 +27,22 @@ def cross_flow_heat_exchanger_geometry_setup(HEX):
...
+    HEX.design_heat_removed                               = 75000  # add loop
```
Comment (ssankalp26) [inline RCAIDE/Library/Methods/Thermal_Management/Heat_Exchangers/Cross_Flow_Heat_Exchanger/cross_flow_heat_exchanger_geometry_setup.py:39]: add loop

File: RCAIDE/Library/Mission/Common/Initialize/energy.py [hunk 21]
```diff
@@ -61,3 +61,5 @@ def energy(segment):
                         fuel_tank_conditions.mass[:,0]   = segment.analyses.energy.vehicle.networks[network.tag].fuel_lines[fuel_line.tag].fuel_tanks[fuel_tank.tag].fuel.mass_properties.mass
             
                     
+def run_append_battery_segment_conditions(battery, bus, conditions, segment):
+    battery.append_battery_segment_conditions(bus, conditions, segment)
```
Comment (ssankalp26) [inline RCAIDE/Library/Mission/Common/Initialize/energy.py:65]: delete

File: RCAIDE/Library/Components/Energy/Distributors/Electrical_Bus.py [hunk 22]
```diff
@@ -28,15 +30,30 @@ def __defaults__(self):
...
+        
+    def append_operating_conditions(self, segment):
+        append_bus_conditions(self, segment)
+        return
+    
+    def compute_distributor_conditions(self,state):
+        compute_bus_conditions(self, state)
+        return 
+        
+
```
Comment (ssankalp26) [inline RCAIDE/Library/Components/Energy/Distributors/Electrical_Bus.py:58]: extra space

**Review by mclarke2 — APPROVED** [review]
Good Job

**Pull request merged**